### PR TITLE
Add security context for SpiceDB Operator

### PR DESCRIPTION
**Pull request by jawnsy**
Closes: #237

### CODE REVIEW

**Comment by jawnsy** [comment]
recheck

**Review by ecordell — COMMENTED** [review]
Thanks for doing this! Left a couple of questions if you have a second.

File: config/operator.yaml
Comment (jawnsy) [reply config/operator.yaml:75]: I had a failure without it:

```
$ k get pod
NAME                                READY   STATUS             RESTARTS      AGE
spicedb-operator-5b666bff7-kfbtt    1/1     Running            0             39m
spicedb-operator-7d9bf5b5b7-8vv8t   0/1     CrashLoopBackOff   6 (37s ago)   6m30s
```

I don't have the logs anymore but here's what I saw as captured by Cloud Logging:

image

Here's the full message:

```json
{
  "textPayload": "panic: mkdir /tmp/kubebuilder-envtest: read-only file system\n\ngoroutine 1 [running]:\nsigs.k8s.io/controller-runtime/pkg/internal/testing/addr.init.0()\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.1/pkg/internal/testing/addr/manager.go:57 +0x165",
  "insertId": "2walpd5izxoo3b2x",
  "resource": {
    "type": "k8s_container",
    "labels": {
      "cluster_name": "cloud2-dev-green",
      "location": "us-east1",
      "pod_name": "spicedb-operator-7d9bf5b5b7-8vv8t",
      "container_name": "spicedb-operator",
      "project_id": "prefect-dev-cloud2",
      "namespace_name": "spicedb-operator"
    }
  },
  "timestamp": "2023-08-16T20:47:11.609563626Z",
  "severity": "ERROR",
  "labels": {
    "k8s-pod/app": "spicedb-operator",
    "k8s-pod/app_kubernetes_io/instance": "spicedb-operator",
    "k8s-pod/pod-template-hash": "7d9bf5b5b7",
    "compute.googleapis.com/resource_name": "gke-cloud2-dev-green-application-3cc5-79a27aea-bj2k"
  },
  "logName": "projects/prefect-dev-cloud2/logs/stderr",
  "receiveTimestamp": "2023-08-16T20:47:33.420653682Z"
}
```

File: config/operator.yaml
Comment (jawnsy) [reply config/operator.yaml:83]: Kinda! You can set a cluster-wide default, but by default it's still unconfined.

This is the feature you're thinking of: https://kubernetes.io/docs/tutorials/security/seccomp/#enable-the-use-of-runtimedefault-as-the-default-seccomp-profile-for-all-workloads

However, if this flag is not enabled, then you'll still get the previous behavior (Unconfined) so it's best to be explicit so that the behavior is consistent across clusters.

File: config/operator.yaml
Comment (ecordell) [reply config/operator.yaml:75]: Hm, envtest shouldn't be running at all outside of the e2e tests. I think we should look at this a bit more.

File: config/operator.yaml
Comment (jawnsy) [reply config/operator.yaml:79]: I set them in both because the pod context are the defaults that get applied when there are sidecar containers injected with admission controllers (unless the sidecar overrides those). So this is a kind of belt-and-suspenders approach to cover all of our bases

File: config/operator.yaml
Comment (ecordell) [reply config/operator.yaml:75]: I'm guessing it's this: https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/internal/testing/addr/manager.go#L44-L59

Comment (jawnsy) [reply config/operator.yaml:75]: lmaooo good 'ol global init functions.

well, in the interest of not letting the best be the enemy of the good, I'd propose we either:

* Merge with the /tmp volume; or,
* Remove the readOnlyRootFilesystem restriction

to unblock this change, then explore whatever controller-runtime is doing with the upstream maintainers

Comment (jawnsy) [reply config/operator.yaml:75]: One downside of the /tmp volume is that it can cause issues with Cluster Autoscaler, since CA won't evict pods that have volumes attached

https://github.com/kubernetes/autoscaler/issues/3984

File: config/operator.yaml
Comment (ecordell) [reply config/operator.yaml:79]: That's reasonable, thanks

File: config/operator.yaml
Comment (ecordell) [reply config/operator.yaml:75]: I have https://github.com/authzed/controller-idioms/pull/40 which removes controller-runtime from controller-idioms, which I will then pull into https://github.com/authzed/spicedb-operator/pull/243 this PR which removes the spicedb-operator dependency.

Once we get those in, there should be no need for a filesystem anymore 🎉

Comment (jawnsy) [reply config/operator.yaml:75]: What's the best way for us to proceed? I can update this to remove the volumes, but I have no easy way of testing

You should also be able to push to my branch

Comment (ecordell) [reply config/operator.yaml:75]: Can you rebase this branch on `main` and remove the tmp directory? I tested it in a local kind cluster and it seems to work fine.

I can do those steps for you if you prefer, but I like keeping author/committer the same when possible 😄

Comment (jawnsy) [reply config/operator.yaml]: Rebased and removed!

**Review by ecordell — APPROVED** [review]
LGTM

**Pull request merged**
